test/NuGet.Core.Tests/NuGet.Protocol.Tests/ServiceIndexResourceV3Tests.cs

```diff
@@ -24,6 +26,79 @@ public void Constructor_InitializesProperties()
             Assert.Equal("http://unit.test/b", resource.Entries[0].Uri.ToString());
         }
 
+        [Fact]
+        public void GetServiceEntries_InvokesDiagnosticEventForSourceResources()
+        {
+            // Arrange
+            int eventInvokeCount = 0;
+            List<ProtocolDiagnosticServiceIndexEntryEvent> capturedEvents = new List<ProtocolDiagnosticServiceIndexEntryEvent>();
+
+            ProtocolDiagnostics.ServiceIndexEntryEvent += (pdEvent) =>
+            {
+                eventInvokeCount++;
+                capturedEvents.Add(pdEvent);
+            };
+
+            var source = $"https://test/index.json";
+            var content = CreateServiceIndexWithFourResourceTypesTwoHTTP();
+
+            var expectedRequestTime = DateTime.UtcNow;
+            var resource = new ServiceIndexResourceV3(content, expectedRequestTime, new Configuration.PackageSource(source));
+
+            // Act
+            var result = resource.GetServiceEntries(ServiceTypes.SearchQueryService);
+
+            // Assert
+            int httpSourceCapture = 0;
```

Comment:
Source and resource mean very different things in this context, so I suggest being strict with the naming.

```suggestion
            int httpResourceCapture = 0;
```